File: backend/app/models/document.py
from datetime import datetime
from typing import Any
from uuid import UUID, uuid4

from sqlalchemy import DateTime, ForeignKey, Index, String, Text, func, text
from sqlalchemy.dialects.postgresql import JSONB
from sqlalchemy.dialects.postgresql import UUID as PG_UUID
from sqlalchemy.orm import Mapped, mapped_column
from sqlalchemy.types import JSON, Uuid

from app.database.base import Base

PROCESSING_STATUSES = (
    "queued",
    "processing",
    "extracted",
    "pending_review",
    "confirmed",
    "rejected",
    "failed",
)
EXTRACTION_STATUSES = ("not_started", "running", "succeeded", "failed", "skipped")
DOCUMENT_TYPES = ("invoice", "contract", "agreement", "statement", "other")


class Document(Base):
    __tablename__ = "documents"
    __table_args__ = (
        Index(
            "uq_documents_content_hash_hostel_id",
            "content_hash",
            "hostel_id",
            unique=True,
            postgresql_where=text("content_hash IS NOT NULL"),
            sqlite_where=text("content_hash IS NOT NULL"),
        ),
        Index(
            "ix_documents_processing_status_created_at",
            "processing_status",
            "created_at",
        ),
    )

    id: Mapped[UUID] = mapped_column(
        Uuid(as_uuid=True).with_variant(PG_UUID(as_uuid=True), "postgresql"),
        primary_key=True,
        default=uuid4,
    )
    hostel_id: Mapped[UUID] = mapped_column(
        Uuid(as_uuid=True).with_variant(PG_UUID(as_uuid=True), "postgresql"),
        ForeignKey("hostels.id", ondelete="RESTRICT"),
        nullable=False,
        index=True,
    )
    file_name: Mapped[str] = mapped_column(String(255), nullable=False)
    storage_key: Mapped[str] = mapped_column(String(500), nullable=False)
    mime_type: Mapped[str] = mapped_column(String(100), nullable=False)
    document_type: Mapped[str] = mapped_column(String(40), nullable=False, default="invoice")
    processing_status: Mapped[str] = mapped_column(
        String(30),
        nullable=False,
        default="queued",
        server_default="queued",
    )
    extraction_status: Mapped[str] = mapped_column(
        String(30),
        nullable=False,
        default="not_started",
        server_default="not_started",
    )
    extracted_payload: Mapped[dict[str, Any] | None] = mapped_column(
        JSON().with_variant(JSONB(), "postgresql"),
        nullable=True,
    )
    extraction_issues: Mapped[list[dict[str, Any]] | None] = mapped_column(
        JSON().with_variant(JSONB(), "postgresql"),
        nullable=True,
    )
    content_hash: Mapped[str | None] = mapped_column(String(64), nullable=True)
    uploaded_by: Mapped[UUID | None] = mapped_column(
        Uuid(as_uuid=True).with_variant(PG_UUID(as_uuid=True), "postgresql"),
        ForeignKey("users.id", ondelete="SET NULL"),
        nullable=True,
    )
    confirmed_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)
    confirmed_by: Mapped[UUID | None] = mapped_column(
        Uuid(as_uuid=True).with_variant(PG_UUID(as_uuid=True), "postgresql"),
        ForeignKey("users.id", ondelete="SET NULL"),
        nullable=True,
    )
    rejected_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)
    rejected_by: Mapped[UUID | None] = mapped_column(
        Uuid(as_uuid=True).with_variant(PG_UUID(as_uuid=True), "postgresql"),
        ForeignKey("users.id", ondelete="SET NULL"),
        nullable=True,
    )
    rejection_reason: Mapped[str | None] = mapped_column(Text, nullable=True)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        nullable=False,
        server_default=func.now(),
    )

File: frontend/src/types/document.ts
export type ProcessingStatus =
  | 'queued'
  | 'processing'
  | 'extracted'
  | 'pending_review'
  | 'confirmed'
  | 'rejected'
  | 'failed'

export type DocumentType = 'invoice' | 'contract' | 'agreement' | 'statement' | 'other'

export type DocumentRecord = {
  id: string
  hostel_id: string
  file_name: string
  storage_key: string
  mime_type: string
  document_type: DocumentType
  processing_status: ProcessingStatus
  extraction_status: string
  extracted_payload: Record<string, unknown> | null
  extraction_issues: Record<string, unknown>[] | null
  content_hash: string | null
  uploaded_by: string | null
  confirmed_at: string | null
  confirmed_by: string | null
  rejected_at: string | null
  rejected_by: string | null
  rejection_reason: string | null
  created_at: string
}

export type PaginatedDocuments = {
  items: DocumentRecord[]
  page: number
  page_size: number
  total: number
}

export const DOCUMENT_TYPES: DocumentType[] = [
  'invoice',
  'contract',
  'agreement',
  'statement',
  'other',
]

export const ACTIVE_PROCESSING_STATUSES: ProcessingStatus[] = ['queued', 'processing']

export const TERMINAL_PROCESSING_STATUSES: ProcessingStatus[] = [
  'pending_review',
  'confirmed',
  'rejected',
  'failed',
]

export const DOCUMENT_POLL_MS = 3000
